### PR TITLE
Add Unified Focal Loss

### LINKED ISSUE #4413: Unified Focal Loss
**Is your feature request related to a problem? Please describe.**
Having the best loss available.

**Describe the solution you'd like**
Implement the Unified Focal Loss

**Additional context**
The Unified Focal Loss has been shown to be constantly better than quite a few other alternatives that are implemented din MONAI.

Paper: https://www.sciencedirect.com/science/article/pii/S0895611121001750
Original Tensorflow Code: https://github.com/mlyg/unified-focal-loss/blob/main/loss_functions.py

### PR DESCRIPTION
**Pull request by JohnMasoner**
Fixes #4413 

### Description
Add Unified Focal Loss, about the [issue](https://github.com/Project-MONAI/MONAI/issues/4413)

### Status
**Work in progress**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by wyli** [comment]
thanks, please sign the commit https://github.com/Project-MONAI/MONAI/blob/dev/CONTRIBUTING.md#signing-your-work  there are also some minor typing issues https://github.com/Project-MONAI/MONAI/runs/6858724085?check_suite_focus=true#step:7:532

**Comment by wyli** [comment]
Hi @danieltudosiu, could you help review this one if you have some time? thanks!

**Comment by danieltudosiu** [comment]
@wyli will have a look over the weekend.

File: monai/losses/unified_focal_loss.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+
+        # Calculate true positives (tp), false negatives (fn) and false positives (fp)
+        tp = torch.sum(y_true * y_pred, dim=axis)
+        fn = torch.sum(y_true * (1 - y_pred), dim=axis)
+        fp = torch.sum((1 - y_true) * y_pred, dim=axis)
+        dice_class = (tp + self.epsilon) / (tp + self.delta * fn + (1 - self.delta) * fp + self.epsilon)
+
+        # Calculate losses separately for each class, enhancing both classes
+        back_dice = 1 - dice_class[:, 0]
+        fore_dice = (1 - dice_class[:, 1]) * torch.pow(1 - dice_class[:, 1], -self.gamma)
```
Comment (wyli) [inline monai/losses/unified_focal_loss.py:89]: looks like this is for binary segmentation only, so probably `include_background=False` shouldn't be provided as an option and the input must have 2 channels. the docstring should also clarify these.

**Comment by JohnMasoner** [comment]
@wyli Thank you for reviewing the code. I'm thinking about whether to add multiple categories or just support binary segmentation only.

**Comment by wyli** [comment]
thanks, I think the initial support of the binary cases is good (as long as we have the docstring and input check), we can create a new feature request/PR for multiclass extension

File: monai/losses/unified_focal_loss.py [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+
+from monai.networks import one_hot
+from monai.utils import LossReduction
+
+
+class AsymmetricFocalTverskyLoss(_Loss):
+    """
+    AsymmetricFocalTverskyLoss is a variant of FocalTverskyLoss, which attentions to the foreground class.
+
+    Reimplementation of the Asymmetric Focal Tversky Loss (with a build-in sigmoid activation) described in:
```
Comment (wyli) [inline monai/losses/unified_focal_loss.py:26]: perhaps update the docstring here to make it clear it's for binary

Comment (wyli) [inline monai/losses/unified_focal_loss.py:26]: the sigmoid activation -- looks like a typo?

File: monai/losses/unified_focal_loss.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+        self,
+        to_onehot_y: bool = False,
+        delta: float = 0.7,
+        gamma: float = 0.75,
+        epsilon: float = 1e-7,
+        reduction: Union[LossReduction, str] = LossReduction.MEAN,
+    ) -> None:
+        """
+        Args:
+            to_onehot_y (bool, optional): whether to convert `y` into the one-hot format. Defaults to False.
```
Comment (wyli) [inline monai/losses/unified_focal_loss.py:42]:
```suggestion
            to_onehot_y: whether to convert `y` into the one-hot format. Defaults to False.

```
there's no need to include additional typing in the argument, please refer to the other modules as an example such as https://github.com/Project-MONAI/MONAI/blob/669bddf581201f994d1bcc0cb780854901605d9b/monai/losses/dice.py#L59-L66

**Review by wyli — COMMENTED** [review]
thanks, I only have a few minor comments about the documentation, otherwise we can merge these new modules.

**Comment by JohnMasoner** [comment]
@wyli thanks, you took the time to review my code. this has been a great help to me.

File: monai/losses/unified_focal_loss.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+        self.weight: float = weight
+        self.asy_focal_loss = AsymmetricFocalLoss(gamma=self.gamma, delta=self.delta)
+        self.asy_focal_tversky_loss = AsymmetricFocalTverskyLoss(gamma=self.gamma, delta=self.delta)
+
+    # TODO: Implement this  function to support multiple classes segmentation
+    def forward(self, y_pred: torch.Tensor, y_true: torch.Tensor) -> torch.Tensor:
+        """
+        Args:
+            y_pred : the shape should be BNH[WD], where N is the number of classes.
+                It only support binary segmentation.
```
Comment (wyli) [inline monai/losses/unified_focal_loss.py:195]:
```suggestion
                It only supports binary segmentation.

```

File: monai/losses/unified_focal_loss.py [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+                The input should be the original logits since it will be transformed by
+                    a sigmoid in the forward function.
+            y_true : the shape should be BNH[WD], where N is the number of classes.
+                It only support binary segmentation.
```
Comment (wyli) [inline monai/losses/unified_focal_loss.py:199]:
```suggestion
                It only supports binary segmentation.

```

File: monai/losses/unified_focal_loss.py [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+    Reimplementation of the Asymmetric Unified Focal Tversky Loss described in:
+
+    - "Unified Focal Loss: Generalising Dice and Cross Entropy-based Losses to Handle Class Imbalanced Medical Image Segmentation",
+    Michael Yeung, Computerized Medical Imaging and Graphics
+    """
+
+    def __init__(
+        self,
+        to_onehot_y: bool = False,
+        num_classes: int = 1,
```
Comment (wyli) [inline monai/losses/unified_focal_loss.py:159]: docstring missing for `num_class`, perhaps adding that ``num_classes > 1`` is not supported?

File: monai/losses/unified_focal_loss.py [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+        weight: float = 0.5,
+        gamma: float = 0.5,
+        delta: float = 0.7,
+        reduction: Union[LossReduction, str] = LossReduction.MEAN,
+    ):
+        """
+        Args:
+            to_onehot_y : whether to convert `y` into the one-hot format. Defaults to False.
+            delta : weight of the backgrand. Defaults to 0.7.
```
Comment (wyli) [inline monai/losses/unified_focal_loss.py:168]:
```suggestion
            delta : weight of the background. Defaults to 0.7.

```

File: monai/losses/unified_focal_loss.py [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+            gamma : value of the exponent gamma in the definition of the Focal loss  . Defaults to 0.75.
+            epsilon : it's define a very small number each time. simmily smooth value. Defaults to 1e-7.
```
Comment (wyli) [inline monai/losses/unified_focal_loss.py:170]: what is "simmily smooth value"?

```suggestion
            epsilon : it defines a very small number each time. simmily smooth value. Defaults to 1e-7.

```

File: monai/losses/unified_focal_loss.py [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+        weight: float = 0.5,
+        gamma: float = 0.5,
+        delta: float = 0.7,
+        reduction: Union[LossReduction, str] = LossReduction.MEAN,
+    ):
+        """
+        Args:
+            to_onehot_y : whether to convert `y` into the one-hot format. Defaults to False.
+            delta : weight of the backgrand. Defaults to 0.7.
+            gamma : value of the exponent gamma in the definition of the Focal loss  . Defaults to 0.75.
```
Comment (wyli) [inline monai/losses/unified_focal_loss.py:169]:
```suggestion
            gamma : value of the exponent gamma in the definition of the Focal loss. Defaults to 0.75.

```

File: monai/losses/unified_focal_loss.py [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+        to_onehot_y: bool = False,
+        delta: float = 0.7,
+        gamma: float = 2,
+        epsilon: float = 1e-7,
+        reduction: Union[LossReduction, str] = LossReduction.MEAN,
+    ):
+        """
+        Args:
+            to_onehot_y : whether to convert `y` into the one-hot format. Defaults to False.
+            delta : weight of the backgrand. Defaults to 0.7.
```
Comment (wyli) [inline monai/losses/unified_focal_loss.py:109]:
```suggestion
            delta : weight of the background. Defaults to 0.7.

```

File: monai/losses/unified_focal_loss.py [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+            gamma : value of the exponent gamma in the definition of the Focal loss  . Defaults to 0.75.
+            epsilon : it's define a very small number each time. simmily smooth value. Defaults to 1e-7.
```
Comment (wyli) [inline monai/losses/unified_focal_loss.py:111]:
```suggestion
            epsilon : it defines a very small number each time. simmily smooth value. Defaults to 1e-7.

```

File: monai/losses/unified_focal_loss.py [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+        to_onehot_y: bool = False,
+        delta: float = 0.7,
+        gamma: float = 0.75,
+        epsilon: float = 1e-7,
+        reduction: Union[LossReduction, str] = LossReduction.MEAN,
+    ) -> None:
+        """
+        Args:
+            to_onehot_y: whether to convert `y` into the one-hot format. Defaults to False.
+            delta : weight of the backgrand. Defaults to 0.7.
```
Comment (wyli) [inline monai/losses/unified_focal_loss.py:45]:
```suggestion
            delta : weight of the background. Defaults to 0.7.

```

File: monai/losses/unified_focal_loss.py [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+            gamma : value of the exponent gamma in the definition of the Focal loss  . Defaults to 0.75.
+            epsilon : it's define a very small number each time. simmily smooth value. Defaults to 1e-7.
```
Comment (wyli) [inline monai/losses/unified_focal_loss.py:47]:
```suggestion
            epsilon : it defines a very small number each time. simmily smooth value. Defaults to 1e-7.

```

**Review by wyli — COMMENTED** [review]
thanks, the tests look good, please help fix the typos then I'll merge this one.

**Review by wyli — APPROVED** [review]
thanks!
